src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  const cloneState = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        for (const data in extraData) {
+          cloneState[data] = extraData[data];
+        }
```

Comment:
You have a method to do this without loop https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign